src/Core/ProjectFilesWatcher.cs

```diff
@@ -0,0 +1,203 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; see the file COPYING.  If not, write to
+ * the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
+ */
+#endregion
+
+using Reko.Core.Scripts;
+using Reko.Core.Services;
+using System;
+using System.Collections;
+using System.Collections.Concurrent;
+using System.Collections.Specialized;
+using System.IO;
+using System.Linq;
+using System.Timers;
+
+namespace Reko.Core
+{
+    using FileWatchersMap = ConcurrentDictionary<string, FileSystemWatcher>;
+
+    /// <summary>
+    /// Watch project files, reload them if they were changed.
+    /// </summary>
+    public class ProjectFilesWatcher
+    {
+        private readonly IDecompilerService decompilerSvc;
+        private readonly DecompilerEventListener eventListener;
+        private readonly ILoader loader;
+        private readonly FileWatchersMap scriptWatchers;
+        private readonly ConcurrentDictionary<string, bool> changedFiles;
+        private IDecompiler? decompiler;
+        private Project? project;
+
+        public ProjectFilesWatcher(IServiceProvider services)
+        {
+            this.decompilerSvc = services.RequireService<IDecompilerService>();
+            this.eventListener = services.RequireService<DecompilerEventListener>();
+            this.loader = services.RequireService<ILoader>();
+            this.Decompiler = decompilerSvc.Decompiler;
+            decompilerSvc.DecompilerChanged += OnDecompilerChanged;
+            this.scriptWatchers = new FileWatchersMap();
+            this.changedFiles = new ConcurrentDictionary<string, bool>();
+        }
+
+        public IDecompiler? Decompiler
+        {
+            get => decompiler;
+            set
+            {
+                if (decompiler != null)
+                {
+                    decompiler.ProjectChanged -= OnProjectChanged;
+                }
+                decompiler = value;
+                if (value != null)
+                {
+                    value.ProjectChanged += OnProjectChanged;
+                }
+                this.Project = Decompiler?.Project;
+            }
+        }
+
+        public Project? Project
+        {
+            get => project;
+            set
+            {
+                if (project != null)
+                {
+                    RemoveScripts(project.ScriptFiles.ToList());
+                    project.ScriptFiles.CollectionChanged -= OnScriptsChanged;
+                }
+                project = value;
+                if (value != null)
+                {
+                    AddScripts(value.ScriptFiles.ToList());
+                    value.ScriptFiles.CollectionChanged += OnScriptsChanged;
+                }
+            }
+        }
+
+        private void AddScripts(IList newScripts)
```

Comment:
Is there any reason why this is not `IEnumerable<ScriptFile>`?

Reply:
`e.NewItems` is `IList` at https://github.com/uxmal/reko/blob/1b160dd5d0ce2adc2790c3826b703bb911c78b0b/src/Core/ProjectFilesWatcher.cs#L186

Reply:
Should I convert it to `AddScripts(e.NewItems.OfType<ScriptFile>()); `?

Reply:
Done in 140746e